card/lib/card/set/format/haml_views.rb

```diff
@@ -12,57 +13,52 @@ module Format
       #   view :my_view, template: :haml  # renders mod/core/view/type/basic/my_view.haml
       #
       #   view :with_instance_variables, template: :haml do
-      #     @hello = "haml"
+      #     @actor = "Mark Haml"
       #   end
       #
       #   # mod/core/view/type/basic/with_instance_variables.haml
-      #   Hello
-      #     = hello
+      #   Luke is played by
+      #     = actor
       #
-      #   > render :with_instance_variables  # => "Hello haml"
+      #   > render :with_instance_variables  # => "Luke is played by Mark Haml"
       module HamlViews
         def haml_view_block view, &block
-          template_path = haml_template_path view
-          if block_given?
-            haml_template_render_block_with_locals view, template_path, &block
-          else
-            haml_template_render_block view, template_path
-          end
+          path = haml_template_path view
+          haml_template_proc ::File.read(path), path, &block
         end
 
-        def haml_template_render_block view, template_path
-          template = ::File.read template_path
+        def haml_template_proc template, path, &block
+          block_locals = block_given?
           proc do |view_args|
-            voo = View.new(self, view, view_args, @voo)
-            with_voo voo do
-              haml_to_html template, view_args, nil, path: template_path
+            with_template_path path do
+              locals = block_locals ? haml_block_locals(view_args, &block) : view_args
+              haml_to_html template, locals, nil, path: path
```

Comment:
I should point out that I removed all three `with_voo` calls from the haml processing in this refactor. They seemed incorrect to me, because the haml blocks should only be called by final_render, which is already within the right with_voo block, so these with_voos were adding a superfluous voo object.

But I often get into trouble by removing things when I don't understand their purpose, so I should get your blessing on this one ;)

Reply:
The only reason I can think of why I added this is that somewhere it wasn't working without it. 
Maybe there are `render_haml` calls somewhere instead of a `render :view` call?

Reply:
If it works with wikirate I'm willing to take the risk, since I know now where I have to look if I encounter render issues with haml.